[Chen Yizhong] iP

yizhong187 opened the pull request:
## 👻 `CasperChat`

> "It’s a dog eat dog world" - Casper

### Why CasperChat is great:
- 👻 *Casper*: Casper is a ghost.
- 👻 **Chat**: You can chat with Casper.

### Casper can count!
1. One ☝️
2. Three 🤟
3. Four 🖖

### When to chat with Casper:
- [x] Morning
- [x] Afternoon
- [x] Evening
- [x] Night
- [ ] While sleeping

```python
def hiCasper():
    print("hello Casper!")
```

You can google more about Casper [here](https://www.google.com/)!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

yizhong187 committed:
Renamed chatbot to Casper

yizhong187 committed:
added functionality of echoing commands

yizhong187 committed:
added functionality of storing tasks and returning

yizhong187 committed:
added printLine method to replace repeated printing of breaking line

yizhong187 committed:
updated to properly use printLine method

yizhong187 committed:
initial commmit

yizhong187 committed:
added the ability to mark tasks as done and changing it back to undone

yizhong187 committed:
initial commit

yizhong187 committed:
added support for tracking of three types of tasks: ToDos, Deadlines and Events

yizhong187 committed:
added test case

yizhong187 committed:
initial commit

yizhong187 committed:
added error handling with custom exceptions as well as abstracted main logic to CommandProcessor class

yizhong187 committed:
added parseArgument method to parse arguments recieved from input

yizhong187 committed:
added delete command

yizhong187 committed:
updated test case

yizhong187 committed:
added ascii art

yizhong187 committed:
updated expected result for test

yizhong187 committed:
edited README for the project

yizhong187 committed:
fixed minor indentation issue

yizhong187 committed:
initial commit

yizhong187 committed:
Move Task and its subclasses into 'tasks' package

yizhong187 committed:
Add 'src/main/java/data/casper.txt'

yizhong187 committed:
Add 'text-ui-test/EXPECTED.TXT'

yizhong187 committed:
Add retrieving of data from local storage

yizhong187 committed:
Update imports to match creation of tasks package

yizhong187 committed:
Replace import wildcards with specific imports

yizhong187 committed:
Add deleteLineInFile function

yizhong187 committed:
Add toSaveString method for saving task to local file

yizhong187 committed:
Add editTaskInSavedTasks, removeTaskFromSavedTasks, addTaskToSavedTasks

yizhong187 committed:
Add handling of CustomIOException from CommandProcessor.processInput(input, tasks)

yizhong187 committed:
Update logic for adding tasks

yizhong187 committed:
Merge branch 'branch-Level-7'

yizhong187 committed:
Initial commit

yizhong187 committed:
Add catch clause for CorruptedSavedTasksException of addSavedTasksToTaskArray(FILE_PATH, tasks)

yizhong187 committed:
Initial commit

yizhong187 committed:
Fix minor details to adhere to code conventions

yizhong187 committed:
Update date-time fields to be LocalDateTime objects instead of Strings

yizhong187 committed:
Merge branch 'branch-Level-8'

yizhong187 committed:
Move Casper into casper package and renamed to CasperChat

yizhong187 committed:
Move CommandProcessor into casper package and renamed to Processor

yizhong187 committed:
Move tasks into casper package

yizhong187 committed:
Initial commit

yizhong187 committed:
Initial commit

yizhong187 committed:
Move exceptions into casper package

yizhong187 committed:
Move PrintUtils to casper package and rename to Ui

yizhong187 committed:
Move FileUtils to casper package and rename to Storage

yizhong187 committed:
Move utils package into casper package

yizhong187 committed:
Edit test file

yizhong187 committed:
Move various components into components package

yizhong187 committed:
Update FILE_PATH to be public

yizhong187 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

yizhong187 committed:
Move components into casper package

yizhong187 committed:
Config gradle

yizhong187 committed:
Update dependencies

yizhong187 committed:
Remove parseArgument method that was moved to Parse class

yizhong187 committed:
Initial commit

yizhong187 committed:
Fix minor formatting issues

yizhong187 committed:
Update gradle for shadowJar

yizhong187 committed:
Add JavaDoc

yizhong187 committed:
Merge branch 'branch-A-JavaDoc'

yizhong187 committed:
Update style to comply with coding standard

yizhong187 committed:
Merge branch 'branch-A-CodingStandard'

yizhong187 committed:
Add find command

yizhong187 committed:
Merge branch 'branch-Level-9'

jktang14 commented on src/main/java/casper/CasperChat.java:
```diff
@@ -0,0 +1,51 @@
+package casper;
```
Well done splitting your files into different packages

yizhong187 replied:
Thanks!

jktang14 commented on src/main/java/casper/CasperChat.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.Scanner;
+
+
+public class CasperChat {
```
Perhaps you could add header comments for CasperChat?

yizhong187 replied:
I would be adding header comments soon. Thanks.

jktang14 commented on src/main/java/casper/components/Processor.java:
```diff
@@ -0,0 +1,122 @@
+        String argument = inputParts.length > 1 ? inputParts[1] : null;
+
+        switch (command) {
+            case "list" -> {
```
Should case be on the same line as switch based on the coding standard?

yizhong187 replied:
I'll make the necessary changes. Thanks!

jktang14 commented on src/main/java/casper/tasks/ToDo.java:
```diff
@@ -0,0 +1,49 @@
+ * Represents a "ToDo" task, which is a type of task that only requires a description.
+ * Inherits the basic task functionalities from the {@link Task} class.
+ */
+public class ToDo extends Task{
```
Should a space be added between { and Task?

yizhong187 replied:
I do think so. Thanks for pointing out!

jktang14 commented on src/test/java/casper/components/TaskListTest.java:
```diff
@@ -0,0 +1,57 @@
+import casper.tasks.ToDo;
+import casper.tasks.Deadline;
+import casper.tasks.Event;
+import org.junit.jupiter.api.Test;
```
Perhaps you could move this line to above line 11 for better consistency?

yizhong187 replied:
I'll make the necessary changes. Thanks!

jktang14 reviewed (commented):
LGTM. I felt that your code effectively used inheritance, packaging. For the most part the coding standard was abided by. There are only some parts where you may have missed out, adding these changes would make your code even more perfect. 👍

yizhong187 committed:
Initial commit

yizhong187 committed:
Fix styling violation

yizhong187 committed:
Fix style violation

yizhong187 committed:
Merge branch 'branch-A-CheckStyle'

yizhong187 committed:
Import images representing user and Casper bot

yizhong187 committed:
Update application entry point

yizhong187 committed:
Rename CasperChat.java to Casper.java and break methods into parts

yizhong187 committed:
Initial commit

yizhong187 committed:
Initial commit

yizhong187 committed:
Rename Ui to Stringifier and update methods to return string, not print

yizhong187 committed:
Fix typo

yizhong187 committed:
Add default case to switch

yizhong187 committed:
Update methods to return string rather than print

yizhong187 committed:
Merge branch 'branch-Level-10'

weiseng18 commented on src/main/java/casper/components/TaskList.java:
```diff
@@ -0,0 +1,159 @@
+    /**
+     * Initializes a new empty TaskList.
+     */
+    private ArrayList<Task> tasks;
```
good use of the plural to indicate that this contains multiple tasks

weiseng18 commented on src/main/java/casper/utils/DateTimeUtils.java:
```diff
@@ -0,0 +1,41 @@
+     * @return A {@code LocalDateTime} object representing the provided date and time.
+     * @throws InvalidDateTimeException If the provided string cannot be parsed into a valid date and time.
+     */
+    public static LocalDateTime convertStringToDateTime(String dateTimeString) throws InvalidDateTimeException {
```
method name very clearly indicate what it does, great!

weiseng18 commented on src/main/java/casper/components/Processor.java:
```diff
@@ -0,0 +1,125 @@
+                return "I have unmarked this task:\n" + taskList.getTaskString(taskIndex);
+
+            } else {
+                String toBeDeletedTaskString = taskList.getTaskString(taskIndex);
```
great indication of what the variable might be used for!

weiseng18 reviewed (commented):
method names especially very clearly indicate what they do, good job!

yizhong187 committed:
Update Gradle configuration for JavaFX dependencies

The current Gradle configuration does not include the necessary
dependencies for JavaFX, and the main class entry point is incorrectly
specified. Assertions are also disabled.

Add JavaFX dependencies to the Gradle config file. Fix the main class
entry point to ensure the correct startup class is used. Enable
assertions to improve debugging and testing capabilities.

These changes are made to ensure the project can compile and run JavaFX
applications correctly, and to support better debugging practices
by enabling assertions.

No additional context is needed at this time.

yizhong187 committed:
Add assertions to file handling and task management methods

The current implementation does not include assertions to verify file
existence and task management operations. Adding assertions will help
ensure that the program behaves as expected during debugging.

Update methods to include assertions:
- Verify file existence after loading saved tasks.
- Check file creation and existence after handling file not found exceptions.
- Confirm that the number of lines read from a file matches the list size.
- Ensure task list size changes as expected after adding or removing tasks.

These changes are made to enhance the reliability and debuggability of
the application by catching potential issues early during development.

No additional context is required for these assertions.

yizhong187 committed:
Merge pull request #1 from yizhong187/branch-A-Assertions

Add JavaFX Dependencies and Main Class Configuration; Enable Assertions for Improved Debugging

yizhong187 committed:
Merge pull request #2 from yizhong187/master

Merge master branch onto branch-A-CodeQuality

yizhong187 committed:
Add Javadoc comments to methods

The current codebase lacks detailed Javadoc comments for several
methods, making it difficult to understand their purpose and usage.

Update multiple methods to include Javadoc comments. Describe method
parameters, return values, and any exceptions thrown. Provide clear
explanations of method functionality and usage.

These changes are made to improve code documentation and facilitate
better understanding and maintenance of the codebase.

No additional context is required at this time.

yizhong187 committed:
Merge branch 'master' into branch-A-CodeQuality to pull updated Javadocs

yizhong187 committed:
Update imports

yizhong187 committed:
Merge branch 'master' into branch-A-CodeQuality

yizhong187 committed:
Refactor Processor and abstract command handling

The processInput method in the Processor class handles all command
logic within a single method, making it complex and difficult to
maintain. The code for each command type is mixed together, impacting
readability and testability.

Abstract command handling into dedicated methods: handleBye,
handleList, handleTaskModification, handleFind, and handleTaskCreation.
Update Javadoc to accurately reflect the exceptions thrown by each method.

These changes improve code readability and maintainability by
separating concerns. The refactored processInput method simplifies
the logic and centralizes command handling, facilitating future
modifications and extensions.

No additional context is needed at this time.

yizhong187 committed:
Delete unused code

yizhong187 committed:
Merge pull request #3 from yizhong187/branch-A-CodeQuality

Improve Code Quality by Refactoring Large Methods and Removing Unused Methods

yizhong187 committed:
Removed unused method

yizhong187 committed:
Add comparators for task sorting

The current implementation lacks comparators for sorting tasks by
various criteria.

Add comparators to sort tasks alphabetically, by completion status,
by type, and by due/start dates. Include Javadoc comments to describe
the purpose and functionality of each comparator and their methods.

These changes are made to provide sorting functionality for different
task attributes.

No additional context is required at this time.

yizhong187 committed:
Initial commit for InvalidArgumentException class

The codebase does not have a specific exception for invalid arguments
provided to commands, and it lacks a mechanism to suggest valid
arguments.

Add the `InvalidArgumentException` class to handle cases where an
invalid argument is given for a command. This exception includes a
message with the invalid argument, the command, and a list
of valid arguments as a suggested solution.

These changes are made to improve error handling by providing more
informative and actionable error messages related to invalid command
arguments.

No additional context is required at this time.

yizhong187 committed:
Add various getters

yizhong187 committed:
Add toSaveString method for tasklist serialization

The current implementation lacks a method to convert the entire tasklist
into a string format suitable for saving.

Add the `toSaveString` method to generate a string representation of all
tasks in the list. Each task's representation is separated by a newline
to facilitate saving. The method returns the resulting string, ensuring
that the task list can be easily serialized for storage.

These changes are made to enable the serialization of tasks into a
format suitable for saving, improving the persistence of task data.

No additional context is required at this time.

yizhong187 committed:
Remove debugging print statement

yizhong187 committed:
Add replaceAllSavedTasks method

The current implementation lacks a method to overwrite an entire task
file with new data.

Add the `replaceAllSavedTasks` method to replace all existing tasks in
the specified file with the provided new text. This method will
completely overwrite the file content with the new task data and handle
potential I/O errors by throwing a `CustomIOException`.

These changes are made to facilitate the updating of task files by
replacing their content with new data, ensuring the file accurately
reflects the current state of tasks.

No additional context is required at this time.

yizhong187 committed:
Add support for new "sort" command in Processor

The current implementation lacks support for a new sorting command to
organize tasks based on various criteria.

Enhance the `processInput` method to include a case for the "sort"
command, which handles sorting tasks by method. Introduce the
`handleSort` method to validate arguments and sort tasks based on the
specified method (`alphabetical`, `type`, `time`, `status`). Ensure that
the sorted task list is saved to the storage.

These changes provide more flexible task sorting options and ensure that
tasks are sorted and saved correctly based on user input.

No additional context is required at this time.

yizhong187 committed:
Add `parseSortArgument` method

The current implementation lacks a dedicated method for parsing and
separating sort command arguments into manageable parts.

Introduce the `parseSortArgument` method to split a sort command
argument into two components: the first three characters and the
remaining string. This method handles cases where the argument is
shorter than or equal to three characters by returning the entire
argument as the first part and leaving the second part empty.

This addition facilitates more effective parsing of sort command inputs
and improves argument handling in command processing.

No additional context is required at this time.

yizhong187 committed:
Add sort to list of valid commands

yizhong187 committed:
Merge branch 'branch-C-Sort'

yizhong187 committed:
Update README to be user guide

yizhong187 committed:
Update application window to not be resizable

yizhong187 committed:
Update README

yizhong187 committed:
Update README

yizhong187 committed:
Add Product Screenshot

yizhong187 committed:
Update README

yizhong187 committed:
Update docs/README.md to be user guide

yizhong187 committed:
Add casper.txt in root dir to .gitignore

yizhong187 committed:
Update user guide format

yizhong187 committed:
Rename file path for jar

yizhong187 committed:
Update product screenshot

yizhong187 committed:
Add Casper Chat title to window

yizhong187 committed:
Fix inconsistency in methods resulting in multiple blank lines in saved file

yizhong187 committed:
Update addTaskToSavedTasks call to reflect change in method signature

yizhong187 committed:
Update user guide to include casper.jar link

yizhong187 committed:
Update README to include casper.jar link

yizhong187 committed:
Fix minor typo

yizhong187 committed:
Fix bug for empty casper.txt

yizhong187 committed:
Fix code style

yizhong187 committed:
Update link to jar file
